### PR TITLE
Color conv nppi

### LINKED ISSUE #231: implement more imageconversions supported by nppi
**Is your feature request related to a problem? Please describe.**
current CCNPPI is supporting only fewer conversions , we can support more using nppi commands 

**Describe the solution you'd like**
support all direct conversions done using nppi , if not direct try to support through indirect conversions

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by venkat0907**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #231 

**Description**
added more conversions  which are possible using nppi commands,if direct conversions are not possible supported them through indirect conversions.

**Alternative(s) considered**

currently indirect conversions were done for 2nd order only , can support nth order in future.

**Type**

Type Choose one: (Bug fix | Feature | Documentation | Testing | Other)

**Screenshots (if applicable)**

**Checklist**

- [x] I have read the [Contribution Guidelines](https://github.com/Apra-Labs/ApraPipes/wiki/Contribution-Guidelines)
- [x] I have written Unit Tests
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach

### CODE REVIEW

File: base/src/CCNPPI.cpp [hunk 1]
```diff
@@ -3,38 +3,856 @@
...
-	Detail(CCNPPIProps &_props) : props(_props)
+	Detail(CCNPPIProps& _props) : props(_props)
 	{
 		nppStreamCtx.hStream = props.stream;
 	}
 
-	~Detail()
+	~Detail() {}
+
+	enum Imageformats
```
Comment (mraduldubey) [inline base/src/CCNPPI.cpp:20]: put a comment - this enum has to be same as ImageMetadata enum

File: base/src/CCNPPI.cpp [hunk 2]
```diff
@@ -3,38 +3,856 @@
...
+		UYVY,
+		YUYV,
+		NV12,
+		BAYERBG10 = 20,
+		BAYERBG8,
+		BAYERGB8,
+		BAYERGR8,
+		BAYERRG8
+	};
+
```
Comment (mraduldubey) [inline base/src/CCNPPI.cpp:40]: enumSize = BAYERRG8 + 1 // put comment - last value plus one

File: base/src/CCNPPI.cpp [hunk 3]
```diff
@@ -3,38 +3,856 @@
...
+		{15, 4},  // NV12 to RGB
+		{15, 5}, // NV12 to RGBA
+		{15, 12},  // NV12 to YUV420
+	    {10, 11 }  // yuv411 to YUV444
+	};
+	int convmatrix[25][25][2] = { {-1} };
+
+	void setConvMatrix() {
+
+		for (int i = 0; i < 25; i++) {
```
Comment (mraduldubey) [inline base/src/CCNPPI.cpp:85]: do not hardcode

File: base/src/CCNPPI.cpp [hunk 4]
```diff
@@ -78,130 +896,55 @@ class CCNPPI::Detail
...
-				srcSize[0],
-				nppStreamCtx
-			);
-			if (status != NPP_SUCCESS)
+			if (intermediateFrameType = FrameMetadata::RAW_IMAGE)
 			{
-				LOG_ERROR << "nppiCopy_8u_C1C4R_Ctx failed<" << status << ">";
-				return false;
-			}
+				auto outputRawMetadata = FrameMetadataFactory::downcast<RawImageMetadata>(mIntermediate);
```
Comment (mraduldubey) [inline base/src/CCNPPI.cpp:906]: rename this variable please - this is so confusing

File: base/src/CCNPPI.cpp [hunk 5]
```diff
@@ -78,130 +896,55 @@ class CCNPPI::Detail
...
-			// CUDA MEMCPY Y
-			auto cudaStatus = cudaMemcpy2DAsync(dst[0], dstPitch[0], src[0], srcPitch[0], srcRowSize[0], srcSize[0].height, cudaMemcpyDeviceToDevice, props.stream);
-			// CUDA MEMSET U V
-
-			if (cudaStatus != cudaSuccess)
-			{
-				LOG_ERROR << "copy failed<" << cudaStatus << ">";
-				return false;
-			}
+				auto outputRawMetadata = FrameMetadataFactory::downcast<RawImagePlanarMetadata>(mIntermediate);
```
Comment (mraduldubey) [inline base/src/CCNPPI.cpp:917]: rename variable please

File: base/src/CCNPPI.cpp [hunk 6]
```diff
@@ -78,130 +896,55 @@ class CCNPPI::Detail
...
+				intermediateChannels = outputRawMetadata->getChannels();
 
-			cudaStatus = cudaMemset2DAsync(dst[1],
-				dstPitch[1],
-				128,
-				dstSize[1].width,
-				dstSize[0].height,
-				props.stream
-			); 
+				for (auto i = 0; i < outputChannels; i++)
```
Comment (mraduldubey) [inline base/src/CCNPPI.cpp:921]: same changes - see above comment

File: base/test/ccnppi_tests.cpp [hunk 7]
```diff
@@ -11,423 +11,766 @@
...
-	auto outputPinId = copy2->getAllOutputPinsByType(FrameMetadata::RAW_IMAGE_PLANAR)[0];
+		copy2 = boost::shared_ptr<Module>(new CudaMemCopy(CudaMemCopyProps(cudaMemcpyDeviceToHost, stream)));
+		ccnppi->setNext(copy2);
 
+		sink = boost::shared_ptr<ExternalSinkModule>(new ExternalSinkModule());
+		copy2->setNext(sink);
 
-	auto m3 = boost::shared_ptr<ExternalSinkModule>(new ExternalSinkModule());
-	copy2->setNext(m3);
+		PipeLine p("test");
```
Comment (mraduldubey) [inline base/test/ccnppi_tests.cpp:59]: not required

**Review by mraduldubey — CHANGES_REQUESTED** [review]
some small changes are yet to be made

File: base/include/CCNPPI.h [hunk 8]
```diff
@@ -45,10 +46,12 @@ class CCNPPI : public Module
 	boost::shared_ptr<Detail> mDetail;
 
 	bool mNoChange;
+	bool intermediateConv = false;
```
[the remark this answers is not in the document]
Comment (venkat0907) [reply base/include/CCNPPI.h:49]: removed it added it in the setMetadata

File: base/src/CCNPPI.cpp (hunk 2 above)
Comment (venkat0907) [reply base/src/CCNPPI.cpp:40]: changed it added enumSize

File: base/src/CCNPPI.cpp (hunk 1 above)
Comment (venkat0907) [reply base/src/CCNPPI.cpp:20]: made a comment

File: base/include/CCNPPI.h [hunk 9]
```diff
@@ -45,10 +46,12 @@ class CCNPPI : public Module
...
 	int mInputFrameType;
 	int mOutputFrameType;
 	size_t mFrameLength;
 	framemetadata_sp mOutputMetadata;
+	framemetadata_sp mIntermediateMetadata;
 	std::string mOutputPinId;
 	CCNPPIProps props;
```
[the remark this answers is not in the document]
Comment (venkat0907) [reply base/include/CCNPPI.h:56]: changed it

File: base/src/CCNPPI.cpp (hunk 4 above)
Comment (venkat0907) [reply base/src/CCNPPI.cpp:906]: changed it to intermediateRawMetadata

File: base/src/CCNPPI.cpp (hunk 5 above)
Comment (venkat0907) [reply base/src/CCNPPI.cpp:917]: changed it to intermediateRawPlanarMetadata

File: base/src/CCNPPI.cpp (hunk 6 above)
Comment (venkat0907) [reply base/src/CCNPPI.cpp:921]: changed it to intermediateChannels

File: base/test/ccnppi_tests.cpp (hunk 7 above)
Comment (venkat0907) [reply base/test/ccnppi_tests.cpp:59]: removed it

**Review by mraduldubey — APPROVED** [review]

**Review by mraduldubey — APPROVED** [review]

**Pull request merged**